CPO-13 Dynamic configuration (#336)
hazelcast/hz-docs: docs/modules/configuration/pages/configuring-with-system-properties.adoc

> = Configuring with System Properties and Environment Variables

Author: The configuration override section is meant to just let users know that it is possible. Users could use sys props or environment variables without the need to use a configuration file. This is why it is its own page rather than being embedded.

> = Configuring with System Properties and Environment Variables

Reviewer: Yes I that is also fine. It could have separate page. I just pointed out that we shouldn't have duplicate parts. I can see you fixed that in https://github.com/hazelcast/hz-docs/pull/336/commits/d6508ff6e8b549b97d3e755ab9a1d76cf1974e7b so count this as resolved.

      hazelcast.property.foo: value
    ...
----
--

Java::
+
--
[source,java]
----
Config config = new Config() ;
config.setProperty( "hazelcast.property.foo", "value" );
----
--
JVM::
+
--
.Java command
[source,shell]
----
java -Dhazelcast.property.foo=value
----

.JAVA_OPTS
[source,shell]
----
JAVA_OPTS="-Dhazelcast.property.foo=value"
----
--
====

== Naming System Properties and Environment Variables

The names of system properties and environment variables follow a consistent pattern, where the names match those in the YAML declarative configuration file. See xref:configuring-declaratively.adoc[].

NOTE: All entries need to mirror the YAML configuration structure which differs slightly to XML. And, you cannot configure settings that use YAML lists.

> The names of environment variables follow this pattern:

Author: ```suggestion
The names of environment variables follow these rules:
```

> Environment variables must be named, using the following pattern:

Author: ```suggestion
The names of environment variables follow this pattern:
```

* Each configuration entry needs to start with `HZ_` for member configuration or `HZCLIENT_` for client configuration.
* A new configuration level should be introduced with an underscore (`_`).
* Dashes (`-`) should be removed.
* Names should be in upper case.

> System properties must be named, using the following rules:

Author: ```suggestion
The names of system properties must follow these rules:
```

* A new configuration level should be introduced with an underscore (`_`).
* Dashes (`-`) should be removed.
* Names should be in upper case.

System properties must be named, using the following rules:

* Each configuration entry needs to start with `hz.` for member configuration or `hz-client.` for client configuration.
* A new configuration level should be introduced with a dot (`.`).
* Names should be in lower case.

For example, consider the following configuration settings.

[source,yaml]
----
hazelcast:
  cluster-name: dev
  network:
    port:
      auto-increment: true
      port-count: 100
      port: 5701
    join:
      auto-detection:
        enabled: true
----

You can override these settings by using the following environment variables or system properties:

[source,shell]
----
HZ_CLUSTERNAME=dev
HZ_NETWORK_PORT_AUTOINCREMENT=true
HZ_NETWORK_PORT_PORTCOUNT=100
HZ_NETWORK_PORT_PORT=5701
HZ_NETWORK_JOIN_AUTODETECTION_ENABLED=true
----

[source,shell]
----
hz.cluster-name=dev
hz.network.port.auto-increment=true
hz.network.port.port-count=100
hz.network.port.port=5701

> hz.network.join.auto-detection.enabled=true

Author: This is just explaining the naming. We provide ways to use sys props in the section above. I can add the JAVA opts to that section?

> hz.network.join.auto-detection.enabled=true

Author: I think adding the `JAVA_OPTS` part would be distracting.
